### PR TITLE
Detect using logging before configuration

### PR DESCRIPTION
**Pull request by jasontedor**
It can easily happen that we touch a logger before logging is configured due to chains of static intializers and other such scenarios. This commit adds detection for this mechanism that will fail startup if we touch a logger before logging is configured. This is a bug that will cause builds to fail.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java [hunk 1]
```diff
@@ -51,9 +55,29 @@
...
 import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.StreamSupport;
 
 public class LogConfigurator {
 
+    private static final AtomicBoolean error = new AtomicBoolean();
+
+    private static final StatusListener LISTENER = new StatusConsoleListener(Level.ERROR) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: Maybe name it `UNCONFIGURED_LOG_CATCHER` or something.

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java [hunk 2]
```diff
@@ -51,9 +55,29 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.StreamSupport;
 
 public class LogConfigurator {
 
+    private static final AtomicBoolean error = new AtomicBoolean();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: loggedBeforeConfigured?

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java [hunk 3]
```diff
@@ -79,9 +103,25 @@ public static void configureWithoutConfig(final Settings settings) {
...
+        } finally {
+            StatusLogger.getLogger().removeListener(LISTENER);
+        }
         configure(environment.settings(), environment.configFile(), environment.logsFile());
     }
 
+    private static void checkErrorListener() {
+        assert errorListenerIsRegistered() : "expected error listener to be registered";
+        if (error.get()) {
+            throw new IllegalStateException("status logger logged an error before logging was configured");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: Maybe change the message to something like "used logging before it was configured"?

**Review by nik9000 — APPROVED** [review]
Makes sense to me. I left some recommendations around naming things so it'd be easier for someone skimming the classes to know what was happening.

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: The most likely error that is logged here is that we touched a logger before it's configured, but any error message from the status logger is going to trip this check so I'd rather keep the name general. I'm happy to add a comment that explains this though.

**Comment by nik9000** [comment]
Makes sense to me!

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: I pushed 18d12a3441f86e549cf2355f801b13e0bfe04281.

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: I pushed 18d12a3441f86e549cf2355f801b13e0bfe04281.

File: core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java]: I pushed 18d12a3441f86e549cf2355f801b13e0bfe04281.

**Pull request merged**